### DISCUSSION TITLE
Configuration versionning

### DISCUSSION THREAD

**Discussion by menardorama** [keycloak/keycloak#21041] [Ideas]
We are successfully using Keycloak in my company for several years now.

One of the main disadvantage I see is that several admins are configuring clients and sometimes can result in errors of misconfiguration.

On firewalls like fortinet for exemple, we have a centralized appliance that do a versioning of the modification before they are applied on the firewall.

I think that having such feature on keycloak would be very useful, we would just have to handle database backup to be confident in what updates are done on keycloak.

If there is a mistake, it can be rolledback.

**Comment by christian-2** [comment]
Could you not just export the current client configuration from Keycloak and use the resulting JSON as snapshot (plus import it, if rollback should become necessary)? Both import and export can be automated via the [Keycloak REST API](https://www.keycloak.org/docs-api/21.1.2/rest-api/index.html), if I am not mistaken.

**Comment by andi-hader** [comment]
https://github.com/adorsys/keycloak-config-cli
might be healpful
